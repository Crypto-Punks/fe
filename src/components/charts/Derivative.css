.Derivative {
  background-color: #ACDEE3;
  border-radius: .5em;
  width: 90vw;
}
