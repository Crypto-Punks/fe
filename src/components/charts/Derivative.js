import { Line } from 'react-chartjs-2';
import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { getPriceHistory } from '../../services/currencies';
import moment from 'moment';
import styles from './Derivative.css';

const Derivative = ({ id, derivativeDuration }) => {
  const [derivativeArray, setDerivativeArray] = useState([]);
  const [intervals, setIntervals] = useState([]);

  const derivativeOptions = {
    title: {
      display: true,
      text: 'Derivative'
    },
    animation: {
      duration: 1000,
      easing: 'linear',
      showlines: true
    }
  };

  useEffect(() => {
    getPriceHistory(id, derivativeDuration)
      .then(priceHistory => {
        const derivatives = priceHistory.map((element, i) => {
          if(i === 0) return (priceHistory[1 + 1].priceUsd - element.priceUsd) * 100 / (priceHistory[i + 1].time - element.time);
          return (element.priceUsd - priceHistory[i - 1].priceUsd) * 100 / (element.time - priceHistory[i - 1].time);
        });
        setDerivativeArray(derivatives);
        setIntervals(priceHistory.map(item => moment(item.time).format('MMM Do YYYY')));
      });
  }, [derivativeDuration]);
  
  return (
    <div className={styles.Derivative}>
      <Line 
        data={{ 
          labels: intervals, 
          datasets: [{ 
            label: 'rate of change', 
            data: derivativeArray, 
            backgroundColor: 'rgba(255, 0, 0, 0.2)',
            borderWidth: .5,
            borderColor: 'rgba(0, 0, 0, 0.2)' 
          }] 
        }} 
        options={derivativeOptions} /> 
    </div>
  );
};

Derivative.propTypes = {
  id: PropTypes.string.isRequired,
  derivativeDuration: PropTypes.string.isRequired
};

export default Derivative;
