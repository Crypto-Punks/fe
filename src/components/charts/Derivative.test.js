import React from 'react';
import { shallow } from 'enzyme';
import Derivative from './Derivative';

const data = {
  id: 'test',
  derivativeDuration: 'test'
};

describe('Derivative component', () => {
  it('renders Derivative', () => {
    const wrapper = shallow(<Derivative { ...data} />);
    expect(wrapper).toMatchSnapshot();
  });
});
