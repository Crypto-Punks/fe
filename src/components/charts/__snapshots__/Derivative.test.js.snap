// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Derivative component renders Derivative 1`] = `
<div
  className="Derivative"
>
  <Line
    data={
      Object {
        "datasets": Array [
          Object {
            "backgroundColor": "rgba(255, 0, 0, 0.2)",
            "borderColor": "rgba(0, 0, 0, 0.2)",
            "borderWidth": 0.5,
            "data": Array [],
            "label": "rate of change",
          },
        ],
        "labels": Array [],
      }
    }
    options={
      Object {
        "animation": Object {
          "duration": 1000,
          "easing": "linear",
          "showlines": true,
        },
        "title": Object {
          "display": true,
          "text": "Derivative",
        },
      }
    }
  />
</div>
`;
